src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,297 @@
+                    {
+                        globalProperties = getGlobalPropertiesMethod.Invoke(BuildEngine, parameters: null) as IReadOnlyDictionary<string, string>;
+                    }
+                    catch (Exception)
+                    {
+                        // Ignored
+                    }
+                }
+            }
+#endif
+            Dictionary<string, string> newGlobalProperties = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            if (globalProperties != null)
+            {
+                foreach (KeyValuePair<string, string> item in globalProperties)
+                {
+                    newGlobalProperties[item.Key] = item.Value;
+                }
+            }
+
+            newGlobalProperties["ExcludeRestorePackageImports"] = bool.TrueString;
+            newGlobalProperties["OriginalMSBuildStartupDirectory"] = MSBuildStartupDirectory;
+
+            if (IsSolutionPathDefined)
+            {
+                newGlobalProperties["SolutionPath"] = SolutionPath;
+            }
+
+            return newGlobalProperties;
+        }
+
+        /// <summary>
+        /// Gets the file name of the process.
+        /// </summary>
+        /// <returns>The full path to the file for the process.</returns>
+        internal string GetProcessFileName(string processFileName)
+        {
+            if (!string.IsNullOrEmpty(processFileName))
+            {
+                return Path.GetFullPath(processFileName);
+            }
+#if IS_CORECLR
+            // In .NET Core, the path to dotnet is the file to run
+            return Path.GetFullPath(Path.Combine(MSBuildBinPath, "..", "..", "dotnet"));
+#else
+            return Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.exe"));
+#endif
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            _cancellationTokenSource.Dispose();
+        }
+
+        protected virtual IEnumerable<KeyValuePair<string, string>> GetOptions()
+        {
+            yield return new KeyValuePair<string, string>(nameof(Recursive), Recursive.ToString());
+        }
+
+        internal void WriteResponseFile(string responseFilePath)
+        {
+            Log.LogMessage(MessageImportance.Low, "Writing response file '{0}'", responseFilePath);
+
+            try
+            {
+                var responseFile = new StaticGraphRestoreArguments
+                {
+                    EntryProjectFilePath = IsSolutionPathDefined ? SolutionPath : ProjectFullPath,
+                    GlobalProperties = GetGlobalProperties(),
+#if IS_CORECLR
+                    MSBuildExeFilePath = Path.Combine(MSBuildBinPath, "MSBuild.dll"),
+#else
+                    MSBuildExeFilePath = Path.Combine(MSBuildBinPath, "MSBuild.exe"),
+#endif
+                    Options = GetOptions().ToDictionary(i => i.Key, i => i.Value, StringComparer.OrdinalIgnoreCase),
```

Comment:
FYI, I was curious since `GetOptions` uses `yield return`. In fact, both "real" classes override `GetOptions`, but then call `foreach (... in base.GetOptions())`, which increases the number of alocations and state machines the compiler generates. So, I wrote this little benchmark:

<details>
<summary>code</summary>
<p>

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

[MemoryDiagnoser]
public class Program
{
    private RealClass _realClass = new RealClass();

    internal static void Main()
    {
        BenchmarkRunner.Run<Program>();
    }

    [Benchmark]
    public Dictionary<string, string> FromIEnumerable()
    {
        return _realClass.GetOptions().ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
    }

    [Benchmark]
    public Dictionary<string, string> NewDictionary()
    {
        return new Dictionary<string, string>()
        {
            ["Recursive"] = true.ToString(),
            ["CleanupAssetsForUnsupportedProjects"] = false.ToString(),
            ["DisableParallel"] = false.ToString(),
            ["Force"] = false.ToString(),
            ["ForceEvaluate"] = false.ToString(),
            ["HideWarningsAndErrors"] = false.ToString(),
            ["IgnoreFailedSources"] = false.ToString(),
            ["Interactive"] = false.ToString(),
            ["NoCache"] = false.ToString(),
            ["RestorePackagesConfig"] = false.ToString()
        };
    }
}

internal class BaseClass
{
    public virtual IEnumerable<KeyValuePair<string, string>> GetOptions()
    {
        yield return new KeyValuePair<string, string>("Recursive", true.ToString());
    }
}

internal class RealClass : BaseClass
{
    public override IEnumerable<KeyValuePair<string, string>> GetOptions()
    {
        foreach (KeyValuePair<string, string> option in base.GetOptions())
        {
            yield return option;
        }

        yield return new KeyValuePair<string, string>("CleanupAssetsForUnsupportedProjects", false.ToString());
        yield return new KeyValuePair<string, string>("DisableParallel", false.ToString());
        yield return new KeyValuePair<string, string>("Force", false.ToString());
        yield return new KeyValuePair<string, string>("ForceEvaluate", false.ToString());
        yield return new KeyValuePair<string, string>("HideWarningsAndErrors", false.ToString());
        yield return new KeyValuePair<string, string>("IgnoreFailedSources", false.ToString());
        yield return new KeyValuePair<string, string>("Interactive", false.ToString());
        yield return new KeyValuePair<string, string>("NoCache", false.ToString());
        yield return new KeyValuePair<string, string>("RestorePackagesConfig", false.ToString());
    }
}
```

</p>
</details>

|          Method |     Mean |   Error |  StdDev |   Gen0 | Allocated |
|---------------- |---------:|--------:|--------:|-------:|----------:|
| FromIEnumerable | 726.8 ns | 8.78 ns | 7.78 ns | 0.2594 |    1088 B |
|   NewDictionary | 436.8 ns | 5.42 ns | 4.80 ns | 0.2370 |     992 B |

so, the perf difference is only about 0.3/1000 of the JSON parsing micro-optimization you made, but it shows that using `yield return` to create lists/dictionaries takes nearly twice as much CPU time as creating the final data structure directly, plus a slight increase in allocations and therefore increase in GC pauses.

Perhaps this isn't enough to warrant changing in a "once per process" method call (such as this one), but I think it's evidence that we should avoid methods returning `IEnumerable` in loops. Anyway, I wanted to share the benchmark results, since I went to the effort of writing it.

Reply:
I'll just create the dictionary and stop using `IEnumerable` for this code path